cas/service.c

```diff
@@ -2181,6 +2185,48 @@ static int cas_exec(struct cas_fom *fom, enum m0_cas_opcode opc,
 		ret = m0_ctg_insert(ctg_op, ctg, &kbuf, &vbuf, next);
 		break;
 	case CTG_OP_COMBINE(CO_DEL, CT_BTREE):
+		op = cas_op(fom0);
+		rec = cas_at(op, rec_pos);
+
+		/*
+		 * Lookup for the value that is going to delete soon.
+		 * Passing -1 for the next phase indicating we don't
+		 * have the next phase yet.
+		 */
+		ret = m0_ctg_lookup(ctg_op, ctg, &kbuf, -1);
+		if (ctg_op->co_rc != 0) {
+			M0_LOG(M0_DEBUG, "Failed lookup on delete. rc=%d",
+				ctg_op->co_rc);
+			m0_fom_phase_set(fom0, next);
+			break;
+		}
+		m0_ctg_lookup_result(ctg_op, &lbuf);
+
+		/*
+		 * Copy the lookup result and the key to m0_cas_op buffers.
+		 * We need this because del operation in the balanaced tree
+		 * will rebalance the tree that makes the looked-up value
+		 * not avaialble afterwards.
+		 *
+		 * Since m0_cas_op is completely added to the fol rec, we
+		 * have these buffers available for the FDMI and don't have
+		 * to send them to the client.
+		 */
+
+		/* Both buffers released in cas_fom_fini(). */
+		m0_rpc_at_init(&rec->cr_key);
+		rec->cr_key.ab_type = M0_RPC_AT_INLINE;
+		rec->cr_key.u.ab_buf = M0_BUF_INIT0;
+		m0_buf_copy(&rec->cr_key.u.ab_buf, &kbuf);
```

Comment:
Is the "key" of DEL operation already included in FDMI/FOL record?
Do we have to do it again?
Or is it going to be processed in m0_ctg_delete()?

Reply:
I did not find the place it was added to the del. It may happen it was not because del event did not come at all. I will check this.

Reply:
PS. Checked. Nothing. There is send key phase but it is part of another state machine phases sequence and is not involved in the del record route.